lib/instrumentation/modules/http2.js

```diff
@@ -0,0 +1,86 @@
+'use strict'
+
+var debug = require('debug')('elastic-apm')
+var eos = require('end-of-stream')
+var shimmer = require('../shimmer')
+
+module.exports = function (http2, agent) {
+  debug('shimming http2.createServer function')
+  shimmer.wrap(http2, 'createServer', wrapCreateServer)
+  shimmer.wrap(http2, 'createSecureServer', wrapCreateServer)
+
+  return http2
+
+  // The `createServer` function will unpatch itself after patching
+  // the first server prototype it patches.
+  function wrapCreateServer (original) {
+    return function wrappedCreateServer (options, handler) {
+      var server = original.apply(this, arguments)
+      shimmer.wrap(server.constructor.prototype, 'emit', wrapEmit)
+      wrappedCreateServer.__obUnwrap()
+      return server
+    }
+  }
+
+  function wrapEmit (original) {
+    var patched = false
+    return function wrappedEmit (event, stream, headers) {
+      if (event === 'stream') {
+        if (!patched) {
+          patched = true
+          shimmer.wrap(stream.constructor.prototype, 'respond', wrapRespond)
+          shimmer.wrap(stream.constructor.prototype, 'end', wrapEnd)
+        }
+
+        // TODO: Figure out if we need to build a response object too
+        var request = {
+          headers,
+          socket: stream.session.socket,
+          method: headers[':method'],
+          url: headers[':path'],
+          httpVersion: '2.0'
+        }
+
+        debug('intercepted request event call to http2.Server.prototype.emit')
+
+        var trans = agent.startTransaction()
+        trans.type = 'http2-request'
+        trans.req = request
+        trans.res = {
+          statusCode: 200,
+          headersSent: false,
+          finished: false,
+          headers: null
+        }
+
+        eos(stream, function () {
```

Comment:
@Qard Will the stream not be reused and hence not end across multiple requests and responses, or is this an abstraction on top of the TCP stream?

Reply:
Each request within a connection gets its own stream.